src/ir/type-updating.cpp

```diff
@@ -38,44 +39,125 @@ GlobalTypeRewriter::TypeMap GlobalTypeRewriter::rebuildTypes(
   // Find the heap types that are not publicly observable. Even in a closed
   // world scenario, don't modify public types because we assume that they may
   // be reflected on or used for linking. Figure out where each private type
-  // will be located in the builder. Sort the private types so that supertypes
-  // come before their subtypes.
-  Index i = 0;
-  auto privateTypes = ModuleUtils::getPrivateHeapTypes(wasm);
+  // will be located in the builder.
+  //
+  // There are two code paths here: a new one that is used when there are type
+  // indices to preserve and an old one that is used otherwise. The old code
+  // path is kept around to avoid unnecessary changes to test outputs while we
+  // incrementally add --preserve-type-order to tests that could benefit from
+  // it. Once we are done adding --preserve-type-order to tests, we should
+  // remove the old code path here since the new code path is strictly better.
+  if (wasm.typeIndices.size()) {
+    // New code path, currently used only with --preserve-type-order.
+    auto typeInfo = ModuleUtils::collectHeapTypeInfo(
+      wasm,
+      ModuleUtils::TypeInclusion::UsedIRTypes,
+      ModuleUtils::VisibilityHandling::FindVisibility);
+
+    std::unordered_set<HeapType> additionalSet(additionalPrivateTypes.begin(),
+                                               additionalPrivateTypes.end());
+
+    std::vector<std::pair<HeapType, SmallVector<HeapType, 1>>>
+      privateSupertypes;
+    privateSupertypes.reserve(typeInfo.size());
+    for (auto& [type, info] : typeInfo) {
+      if (info.visibility != ModuleUtils::Visibility::Private &&
+          !additionalSet.count(type)) {
+        continue;
+      }
+      privateSupertypes.push_back({type, {}});
+
+      if (auto super = getDeclaredSuperType(type)) {
+        auto it = typeInfo.find(*super);
+        // Record the supertype only if it is among the private types.
+        if ((it != typeInfo.end() &&
+             it->second.visibility == ModuleUtils::Visibility::Private) ||
+            additionalSet.count(*super)) {
+          privateSupertypes.back().second.push_back(*super);
+        }
+      }
+    }
+
+    // Topological sort to have subtypes first. This is the opposite of the
+    // order we need, so the comparison is the opposite of what we ultimately
+    // want.
+    std::vector<HeapType> sorted;
+    if (wasm.typeIndices.empty()) {
```

Comment:
Line 50 checked `wasm.typeIndices.size()`, so this cannot be empty?

Reply:
Indeed, although once we delete the check on line 50 and the old code path, this will become important. If you would prefer, I can make this a TODO instead.

Reply:
If you plan to quickly delete the old code path sgtm as is, but if not maybe a one-line note here could help.

Reply:
I do plan to delete it relatively quickly 👍

Reply:
In fact, it is already deleted in #6902.